tls/s2n_handshake_io.c

```diff
@@ -831,14 +831,14 @@ int s2n_conn_set_handshake_type(struct s2n_connection *conn)
      * Client sent in the ClientHello. */
     if (conn->actual_protocol_version <= S2N_TLS12 && conn->mode == S2N_SERVER && s2n_allowed_to_cache_connection(conn)) {
         int r = s2n_resume_from_cache(conn);
-        if (r == S2N_SUCCESS || (r < 0 && S2N_ERROR_IS_BLOCKING(s2n_errno))) {
+        if (r == S2N_SUCCESS || (r < S2N_SUCCESS && S2N_ERROR_IS_BLOCKING(s2n_errno))) {
```

Comment:
Could this be:
```suggestion
        if (r == S2N_SUCCESS || (r != S2N_SUCCESS && S2N_ERROR_IS_BLOCKING(s2n_errno))) {
```
or is the `<` necessary?

Reply:
I don't see why not here. @lrstewart , any reason why we should do this?

Reply:
It depends on if the function is conflating the error signal with some output value, e.g. byte len, count, etc.

Reply:
If we keep the comparisons, wouldn't it be better to keep the numeric `0`?

Having `< S2N_SUCCESS` (read: less than success) just looks a bit odd, IMO, but please feel free to ignore my comments if this is the s2n convention.

Reply:
I agree it's weird; if i had it my way we wouldn't have any functions that return values other than `-1` or `0` (that's exactly what `S2N_RESULT` is, BTW). Unfortunately, with the `int` type it can return whatever it wants and that can lead to issues when doing a `!= S2N_SUCCESS`. As we move more functions to return `S2N_RESULT` it will be less of an issue.

Reply:
Thanks @camshaft, that makes sense.

Also, after we switch to `S2N_RESULT`, the return value wouldn't be `int` and so we would have to change the `< S2N_SUCCESS` predicate once more?

Reply:
Correct. There are functions to inspect the value of the result:

https://github.com/aws/s2n-tls/blob/dffe5612e9558bdbedd985e1f67b4c3efd6992c3/utils/s2n_result.c#L84-L93

Since results are only supposed to be created with the `S2N_RESULT_OK` and `S2N_RESULT_ERR` we can do an exact comparison.

https://github.com/aws/s2n-tls/blob/dffe5612e9558bdbedd985e1f67b4c3efd6992c3/utils/s2n_result.h#L26-L30

Reply:
also for more context here's the check in `POSIX_GUARD` (>= and not ==)

https://github.com/aws/s2n-tls/blob/dffe5612e9558bdbedd985e1f67b4c3efd6992c3/utils/s2n_safety_macros.h#L346

Reply:
Thank you so much!